_posts/2019-04-01-functional-introduction.md

```diff
@@ -0,0 +1,106 @@
+
+#### Ważniejsze cechy funkcjonalnego podejścia: 
+
+⚙ Higher-order functions - przekazują funkcję jako paramter do innej funkcji - istna incepcja.
+
+Na przykład funkcja `availableCustomers(Supplier<Boolean> customerAvailability)` przyjmuje funkcję jako parametr `Customer::isAvailable`.
+
+Czego wynikiem jest: `availableCustomers(Customer::isAvailable)`
+
+⚙ First-class citizens - czyli traktowanie funkcji jako wartości.
+
+Dla obrazowania przykładu zróbmy sobie funkcję w Kotlinie o wdzięcznej nazwie `adder`: 
+
+*  `fun add(a: Int, b: Int) = (a + b).toDouble()` - nasza funkcja.
+*  `val adder = ::add` - którą przypisujemy do zmiennej.
+*  `val adder: (Int, Int) -> Double = { a, b -> (a + b).toDouble() }` - lub to samo w ten sposób.
+*  `adder(1,1)` - po czym robimy skomplikowane obliczenia.
+
+⚙ Anonymous classes - `() -> "czyli lambdy"`
+
+### Skutki uboczne programowania imperatywnego 🔰
+Impratywnego, czyli takiego z jakim mamy styczność zazwyczaj na początku drogi z programowaniem.
+
+Przykładowo jeśli A == 0 to zwróć B inaczej B++ oraz A--. 
+
+Mamy tutaj mutowalne zmienne, czyli w sumie niszczymy stany obiektów. Jeśli okaże się, że trzeba zmienić wymagania biznesowe to szybko okazuje się, że również zmieniamy wynik działania naszej aplikacji. Programowanie funkcyjne promuje nie-mutowalność obiektów, A.K.A. Immutability. Co jest dobre. Do tego poprzez funkcje ograniczamy interakcje naszego kodu ze światem zewnętrznym. Integracja jest dopiero po skończeniu obliczeń, potem rzucamy wyjątki, zapisujemy do bazy, czy wysyłamy coś po HTTPie. 
+
+### Wracając do wejścia/wyjścia 🚪
+Funkcjny kod ma zapewnić jak najmniejszą ilość efektów ubocznych, czyli mamy `in -> out`, zamiast `in -> file -> exception -> poop -> db -> info -> out`. Takie podejście daje nam bardziej determistyczny sposób działania apki. Również testowanie takiego kodu jest łatwiejsze, bo nie potrzebujemy armii Mocków do wyizolowania przypadku testowego.
+
+### Przechodząc z imperatywnego do funkcyjnego
+//todo: 
+
+
+## Czas na trochę mięsa 🍗
+Kod będzie w Kotlinie. Myślę, że o wiele lepiej oddaje różne idea programowania funkcyjnego bez zbędnego boilercode'u.
+
+Choćby zapis typu funkcji wygląda następująco `(A) -> B`.
+
+Gdzie w Javie byłoby to `Function<? super T, ? extends R>`.
+
+Jak się zapewne domyślasz powyższy kod to dobrze nam znana metoda ` .map { } `
+
+Kotlin bardzo upraszcza kod. Stworzyłem dla ciebie prostą katę, abyś mógł porównać: [Java](https://github.com/braintelligencePL/snippets-and-katas-of-jvm-languages/blob/master/katas/src/main/java/pl/braintelligence/katas/Java_1_SocketsPairs.java) oraz [Kotlin](https://github.com/braintelligencePL/snippets-and-katas-of-jvm-languages/blob/master/katas/src/main/kotlin/pl/braintelligence/katas/Kotlin_1_SocketsPairs.kt), a tutaj [Test Jednostkowy](https://github.com/braintelligencePL/snippets-and-katas-of-jvm-languages/blob/master/katas/src/test/groovy/pl/braintelligence/katas/_1_SocketsPairsTest.groovy).
```

Comment:
IMO fajniej by było jakbyś wrzucił tutaj snippety a nie linkował - od razu by była widoczna różnica

Reply:
+1

Reply:
Te kawałki kodu można nieco skrócić tak, że będą wyglądały na dużo mniejsze:
1. można wyrzucić package i importy - to nie jest konieczne do zrozumienia kodu
2. w kodzie kotlinowym można wyrzucić puste linie w countNumberOfPairs1 - na dobrą sprawę to można by zostawić tylko countNumberOfPairs1 albo countNumberOfPairs2 - zakładam, że nie będziesz opisywał tych implementacji
3. Po dłuższym zastanowieniu dochodzę do wniosku, że ten przykład może być nietrafiony - IMO powinieneś porównywać programowanie funkcyjne z imperatywnym, a tymczasem te kody to porównanie Kotlina z Javą

Reply:
Tak swoją drogą to kod javowy też można bardzo skrócić, tak na szybko wyszło mi coś takiego (pomijam czytelność tego rozwiązania, ale strzelam, że można to jeszcze mocno uprościć):
```java
public static int countNumberOfPairs(int[] socketsColors) {
        return IntStream.of(socketsColors)
                .boxed()
                .collect(Collectors.groupingBy(Integer::intValue))
                .values()
                .stream()
                .mapToInt(list -> list.size() / 2)
                .sum();
    }
```

Reply:
W zamyśle chciałem tylko pokazać te same fragmentu kodu w Kotlinie i Javie bez pokazywania tego we wpisie, bo tak jak mówisz opisujemy programowanie imparatywne, funkcyjne. Jako, że jednak większość zna Javę, a mniejszość Kotlina to zrobiłem prosty przykład, żeby na praktycznym przykładzie pokazać porównanie. Bardziej w zamyśle miałem:  

"Jak ktoś chce to sobie zobaczy (niekoniecznie istotne pod względem wpisu)".